### ISSUE TITLE
tensorflow.js实战

### ISSUE THREAD

**Issue by iloveyou11** [iloveyou11/learning-blog#43]
在这篇文章中，总结使用tensorflow.js做机器学习项目的方法，相比于tensorflow python，前端同学会更容易上手。[tensorflow.js官网地址](https://www.tensorflow.org/js?hl=zh-cn)

### 0. 准备工作
**方法1：html中直接加载tfjs的cdn文件**

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>tensorflow</title>
</head>

<script src="https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@1.0.0/dist/tf.min.js"></script>
<script>
    const a = tf.tensor('1')
    a.print()
</script>

</html>
```

可以看到控制台输出：

```sh
Tensor
    1
```

**方法2：npm安装@tensorflow/tfjs-node包**

```sh
npm i @tensorflow/tfjs-node -S
```

```js
const tf = require("@tensorflow/tfjs-node")
const a = tf.tensor([1, 2])
a.print()
```

输出结果为：

```sh
Tensor
    [1, 2]
```

在以下的全部演示项目中，均采用`parcel`去启动项目，具体流程如下：

1. 全局安装parcel，`npm install -g parcel-bundler`
2. 分别写html文件和js文件，在html中引入对应的js文件
3. 在对应目录下，执行`parcel index.html`，即可启动端口1234，通过`http://localhost:1234`查看页面

### 1. 线性回归

```html
<script src="./script.js"></script>
```

这时，我们收集一组数据x/y，训练线性回归模型，从而可以根据输入值x预测y：

```js
import * as tf from '@tensorflow/tfjs';
import * as tfvis from '@tensorflow/tfjs-vis';

window.onload = async () => {
  // x、y样本数据集
  const xs = [1, 2, 3, 4];
  const ys = [1, 3, 5, 7];

  // 使用 @tensorflow/tfjs-vis 库绘制散点图
  tfvis.render.scatterplot(
    { name: '线性回归训练集' },
    { values: xs.map((x, i) => ({ x, y: ys[i] })) },
    { xAxisDomain: [0, 5], yAxisDomain: [0, 8] }
  );

  // 定义模型。添加含有1个神经元的1层网络
  const model = tf.sequential();
  model.add(tf.layers.dense({ units: 1, inputShape: [1] }));
  model.compile({ loss: tf.losses.meanSquaredError, optimizer: tf.train.sgd(0.1) });

  const inputs = tf.tensor(xs);
  const labels = tf.tensor(ys);

  // 训练模型
  await model.fit(inputs, labels, {
    batchSize: 4,
    epochs: 200,
    callbacks: tfvis.show.fitCallbacks(
      { name: '训练过程' },
      ['loss']
    )
  });

  // 预测
  const output = model.predict(tf.tensor([5]));
  console.log(`如果 x 为 5，那么预测 y 为 ${output.dataSync()[0]}`);
};
```

效果如下所示：

<img alt="线性回归" width="100%" src="https://img.alicdn.com/imgextra/i3/O1CN01Vpumfj1HJaa6JTbAu_!!6000000000737-2-tps-2076-1838.png" />

### 2. 归一化

```html
<script src="./script.js"></script>
```

我们收集了几组身高/体重的数据，并对其进行了归一化处理，以免会因为基数之间的差距而导致模型预测不准。我们将身高、体重数据都压缩到了0-1之间，预测值也是在0-1之间，最后要将预测值进行反归一化，将其转化为真正的体重值。

```js
import * as tf from '@tensorflow/tfjs';
import * as tfvis from '@tensorflow/tfjs-vis';

window.onload = async () => {
  // 样本数据
  const heights = [150, 160, 170];
  const weights = [40, 50, 60];

  // 绘制散点图
  tfvis.render.scatterplot({
    name: '身高体重训练数据'
  }, {
    values: heights.map((x, i) => ({
      x,
      y: weights[i]
    }))
  }, {
    xAxisDomain: [140, 180],
    yAxisDomain: [30, 70]
  });

  // 归一化数据
  const inputs = tf.tensor(heights).sub(150).div(20); //压缩到0-1
  const labels = tf.tensor(weights).sub(40).div(20); //压缩到0-1

  // 构建模型
  const model = tf.sequential();
  model.add(tf.layers.dense({
    units: 1,
    inputShape: [1]
  }));
  model.compile({
    loss: tf.losses.meanSquaredError,
    optimizer: tf.train.sgd(0.1)
  });

  // 训练模型
  await model.fit(inputs, labels, {
    batchSize: 3,
    epochs: 200,
    callbacks: tfvis.show.fitCallbacks({
      name: '训练过程'
    }, ['loss'])
  });

  // 预测 反归一化
  const output = model.predict(tf.tensor([180]).sub(150).div(20));
  console.log(`如果身高为 180cm，那么预测体重为 ${output.mul(20).add(40).dataSync()[0]}kg`);
};
```

<img alt="归一化" width="100%" src="https://img.alicdn.com/imgextra/i1/O1CN01e8tgyA1zAk1StMufM_!!6000000006674-2-tps-1994-1828.png" />

### 3. 逻辑回归

输入x、y，输出预测值

```html
<script src="script.js"></script>
<form onsubmit="predict(this);return false;">
    x: <input type="text" name="x"> 
    y: <input type="text" name="y">
    <button type="submit">预测</button>
</form>
```

我们先写一个函数，产生二分类的平面数据，分为生成以(2,2)为中心label为1的点、以(-2,-2)为中心label为0的点，用于接下来做二分类的任务。

```js
/**
 * @description 二分类数据集生成函数源码
 * @param {*} numSamples 随机生成点的数量
 */
export function getData(numSamples) {
  let points = [];

  function genGauss(cx, cy, label) {
    for (let i = 0; i < numSamples / 2; i++) {
      let x = normalRandom(cx);
      let y = normalRandom(cy);
      points.push({
        x,
        y,
        label
      });
    }
  }

  genGauss(2, 2, 1); //生成以(2,2)为中心，label为1的点
  genGauss(-2, -2, 0); //生成以(-2,-2)为中心，label为0的点
  return points;
}

/**
 * @description 生成正态分布随机点
 * @param mean 均值
 * @param variance 方差，值越大点分布越分散，值越小点分布越密集
 */
function normalRandom(mean = 0, variance = 1) {
  let v1, v2, s;
  do {
    v1 = 2 * Math.random() - 1;
    v2 = 2 * Math.random() - 1;
    s = v1 * v1 + v2 * v2;
  } while (s > 1);

  let result = Math.sqrt(-2 * Math.log(s) / s) * v1;
  return mean + Math.sqrt(variance) * result;
}
```

定义逻辑回归模型进行预测：

```js
import * as tf from '@tensorflow/tfjs';
import * as tfvis from '@tensorflow/tfjs-vis';
import { getData } from './data.js';

window.onload = async () => {
  // 拿到400个数据
  const data = getData(400);

  // 绘制散点图
  tfvis.render.scatterplot({
    name: '逻辑回归训练数据'
  }, {
    values: [
      data.filter(p => p.label === 1),
      data.filter(p => p.label === 0),
    ]
  });

  // 构建模型，主要这里是逻辑回归，使用的损失函数为对数损失 logLoss，可参考http://wiki.fast.ai/index.php/Log_Loss
  const model = tf.sequential();
  model.add(tf.layers.dense({
    units: 1,
    inputShape: [2], //特征数量
    activation: 'sigmoid'
  }));
  model.compile({
    loss: tf.losses.logLoss, //对数损失，逻辑回归不适合使用MSE
    optimizer: tf.train.adam(0.1) //adam优化器，可以自动调节学习率
  });

  const inputs = tf.tensor(data.map(p => [p.x, p.y]));
  const labels = tf.tensor(data.map(p => p.label));

  // 训练模型
  await model.fit(inputs, labels, {
    batchSize: 40,
    epochs: 20,
    callbacks: tfvis.show.fitCallbacks({
      name: '训练效果'
    }, ['loss'])
  });

  // 编写界面输入预测值
  window.predict = (form) => {
    const pred = model.predict(tf.tensor([
      [form.x.value * 1, form.y.value * 1]
    ]));
    console.log(`预测结果：${pred.dataSync()[0]}`);
  };
};
```

<img alt="逻辑回归" width="100%" src="https://img.alicdn.com/imgextra/i3/O1CN01Hz5NpH1nbVmczomsN_!!6000000005108-2-tps-2508-1828.png" />

### 4. XOR

输入x、y，输出预测值

```html
<script src="script.js"></script>
<form action="" onsubmit="predict(this);return false;">
    x: <input type="text" name="x">
    y: <input type="text" name="y">
    <button type="submit">预测</button>
</form>
```

我们通过函数，生成4类数据：
1. 以(2,2)为中心，label=0
2. 以(-2,-2)为中心，label=0
3. 以(-2,2)为中心，label=1
4. 以(2,-2)为中心，label=1

可以看出，x*y>0，则label=0，x*y<0，则label=1。函数如下：

```js
/**
 * @param {*} numSamples 随机生成点的数量
 */
export function getData(numSamples) {
  let points = [];

  function genGauss(cx, cy, label) {
    for (let i = 0; i < numSamples / 2; i++) {
      let x = normalRandom(cx);
      let y = normalRandom(cy);
      points.push({
        x,
        y,
        label
      });
    }
  }

  // 这里产生四类点
  genGauss(2, 2, 0);
  genGauss(-2, -2, 0);
  genGauss(-2, 2, 1);
  genGauss(2, -2, 1);
  return points;
}

/**
 * @description 生成正态分布随机点
 * @param mean 均值
 * @param variance 方差，值越大点分布越分散，值越小点分布越密集
 */
function normalRandom(mean = 0, variance = 1) {
  let v1, v2, s;
  do {
    v1 = 2 * Math.random() - 1;
    v2 = 2 * Math.random() - 1;
    s = v1 * v1 + v2 * v2;
  } while (s > 1);

  let result = Math.sqrt(-2 * Math.log(s) / s) * v1;
  return mean + Math.sqrt(variance) * result;
}
```

接下来，根据以上函数产生的样本数据，开始训练分类模型：

```js
import * as tf from '@tensorflow/tfjs';
import * as tfvis from '@tensorflow/tfjs-vis';
import { getData } from './data.js';

// 不能使用线性逻辑回归
window.onload = async () => {
  const data = getData(400);

  tfvis.render.scatterplot({
    name: 'XOR 训练数据'
  }, {
    values: [
      data.filter(p => p.label === 1),
      data.filter(p => p.label === 0),
    ]
  });

  // 构建多层神经网络
  const model = tf.sequential();
  model.add(tf.layers.dense({
    units: 4,
    inputShape: [2],
    activation: 'relu'
  }));
  model.add(tf.layers.dense({
    units: 1,
    activation: 'sigmoid'
  }));
  model.compile({
    loss: tf.losses.logLoss,
    optimizer: tf.train.adam(0.1)
  });

  const inputs = tf.tensor(data.map(p => [p.x, p.y]));
  const labels = tf.tensor(data.map(p => p.label));

  await model.fit(inputs, labels, {
    epochs: 10,
    callbacks: tfvis.show.fitCallbacks({
      name: '训练效果'
    }, ['loss'])
  });

  window.predict = (form) => {
    const pred = model.predict(tf.tensor([
      [form.x.value * 1, form.y.value * 1]
    ]));
    console.log(`预测结果：${pred.dataSync()[0]}`);
  };
};
```

<img alt="https://img.alicdn.com/imgextra/i3/O1CN01Z1HbAx1pqHmxMIEkN_!!6000000005411-2-tps-2584-1824.png" width="100%" src="" />

### 5. 多分类

这里我们实现一个模型，根据花萼长度、花萼宽度、花瓣长度、花瓣宽度去预测是属于什么类型的鸢尾花。

```html
<script src="script.js"></script>
<form action="" onsubmit="predict(this); return false;">
  花萼长度：<input type="text" name="a"><br>
  花萼宽度：<input type="text" name="b"><br>
  花瓣长度：<input type="text" name="c"><br>
  花瓣宽度：<input type="text" name="d"><br>
  <button type="submit">预测</button>
</form>
```

自定义鸢尾花数据集，这里分为了3类，打标分别为1、2、3，分别代表山鸢尾, 变色鸢尾, 维吉尼亚鸢尾。代码详见[data.js](https://github.com/iloveyou11/my-tfjs/blob/master/project/5%20%E5%A4%9A%E5%88%86%E7%B1%BB/data.js)

获取以上的数据集，训练多分类模型：

```js
import * as tf from '@tensorflow/tfjs';
import * as tfvis from '@tensorflow/tfjs-vis';
import { getIrisData, IRIS_CLASSES } from './data';

window.onload = async () => {
  // 加载数据集和验证集
  const [xTrain, yTrain, xTest, yTest] = getIrisData(0.15); //15%的数据用于验证集

  // 构建模型，这里用到的损失函数为交叉熵损失 categoricalCrossentropy，适用于多分类
  const model = tf.sequential();
  model.add(tf.layers.dense({
    units: 10,
    inputShape: [xTrain.shape[1]],
    activation: 'sigmoid' //也可以使用其他激活函数，如relu
  }));
  model.add(tf.layers.dense({
    units: 3,
    activation: 'softmax' //这里使用softmax激活函数，算出每个类别的概率
  }));
  model.compile({
    loss: 'categoricalCrossentropy', //交叉熵损失，对数损失函数的多分类版本
    optimizer: tf.train.adam(0.1),
    metrics: ['accuracy']
  });

  // 训练模型
  await model.fit(xTrain, yTrain, {
    epochs: 100,
    validationData: [xTest, yTest],
    callbacks: tfvis.show.fitCallbacks({
      name: '训练效果'
    }, ['loss', 'val_loss', 'acc', 'val_acc'], {
      callbacks: ['onEpochEnd']
    })
  });

  window.predict = (form) => {
    const input = tf.tensor([
      [
        form.a.value * 1,
        form.b.value * 1,
        form.c.value * 1,
        form.d.value * 1,
      ]
    ]);
    const pred = model.predict(input);
    console.log(`预测结果：${IRIS_CLASSES[pred.argMax(1).dataSync(0)]}`); //argmax？
  };
};
```

<img alt="多分类" width="100%" src="https://img.alicdn.com/imgextra/i4/O1CN01FS5Gzr1bcAy5aNUQq_!!6000000003485-2-tps-2392-1392.png" />

## 模型互转
**python与JavaScript模型的互转：**
1. 准备工作
`conda activate [name]`激活
`tensorflowjs_converter`检查

2. 开始转换——具体格式建议看文档（github tfjs-converter）
  1）python模型转js模型
  `tensorflowjs_converter --input_format=keras --output_format=tf_layers_model [input_path] [output_path]`
  2）js模型转python模型
  `tensorflowjs_converter --input_format=tf_layers_model --output_format=keras [input_path] [output_path]`

3. 验证模型正确性

## 模型加速
1. 分片(--weight_shared_size_bytes=[size])
 `tensorflowjs_converter --input_format=tf_layers_model --output_format=tf_layers_model --weight_shared_size_bytes=100000 [input_path] [output_path]`
在输出文件夹中会发现很多块分片后的模型，这样在加载模型时可以使用并发实现加速

2. 量化(--quantization_bytes=[n])
 `tensorflowjs_converter --input_format=tf_layers_model --output_format=tf_layers_model --quantization_bytes=2 [input_path] [output_path]`

3. 通过转为tfjs_graph_model来加速模型——内部实现了凸优化(--output_format=tf_graph_model)
  `tensorflowjs_converter --input_format=tf_layers_model --output_format=tf_graph_model [input_path] [output_path]`
